Add full http m3u playlist plus mention mime setup and http headers (#4389)
savonet/liquidsoap: doc/content/playlist_parsers.md

- Xml formats:
  - `video/x-ms-asf`, `audio/x-ms-asx`: [ASX](http://en.wikipedia.org/wiki/Advanced_Stream_Redirector), **strict**
  - `application/smil+xml`, `application/smil+xml`, [SMIL](http://en.wikipedia.org/wiki/Synchronized_Multimedia_Integration_Language), **strict**
  - `application/xspf+xml`, [XSPF](http://en.wikipedia.org/wiki/Xspf), **strict**
  - `application/rss+xml`, [Podcast](http://en.wikipedia.org/wiki/Podcast), **strict**
 

> Playlist format is driven by the **Content-Type** and **Content-Disposition** *(see m3u example below)*. You should make sure that your HTTP endpoint returns appropriate values for those. 

Reviewer: ```suggestion
Playlist format is driven by the **Content-Type** and **Content-Disposition** HTTP headers *(see m3u example below)*. You should make sure that your HTTP endpoint returns appropriate values for those. 
```

  - `application/smil`, `application/smil+xml`, [SMIL](http://en.wikipedia.org/wiki/Synchronized_Multimedia_Integration_Language), **strict**
  - `application/xspf+xml`, [XSPF](http://en.wikipedia.org/wiki/Xspf), **strict**
  - `application/rss+xml`, [Podcast](http://en.wikipedia.org/wiki/Podcast), **strict**
 

> In order for liquidsoap to detect and parse the remote playlist you have to add the MIME type to your settings.http.mime.extnames in the liquidsoap script as well as remote http endpoint has to return correct **Content-Type** and **Content-Disposition** headers *(see m3u example below)*.

Reviewer: ```suggestion
Playlist format is driven by the **Content-Type** and **Content-Disposition** headers *(see m3u example below)*
headers.  You should make sure that your HTTP endpoint returns appropriate values for those. 

As last resort, you should be able to use the `settings.http.mime.extnames` settings to add or adjust support 
for your endpoint's mime-type if liquidsoap supports its corresponding playlist format. See for instance issue [#3451](https://github.com/savonet/liquidsoap/issues/3451).
```

and some programming magic. You can check the code source for `playlist.reloadable` in our standard library
for a detailed example.

### Remote M3U playlist example
Here is an example of a m3u playlist being read from nodejs/express .

liquidsoap script:
```liquidsoap
#!/usr/local/bin/liquidsoap

> settings.http.mime.extnames := [...settings.http.mime.extnames(), ("audio/x-mpegurl", ".m3u")]

Reviewer: ```suggestion
```

p = playlist(reload=10, "http://localhost:8080/radio/playlists/0/playlist.m3u")
```

nodejs/express app:
```js
import express from "express";
const app = express();
app.get('/radio/playlists/:id/playlist.m3u', async (req, res) => {
  const playlist = ["/media/foo.mp3", "/media/bar.mp3"]

>   res.set('Content-Type', 'audio/x-mpegurl')
>   res.set(`Content-Disposition: attachment; filename="playlist-${req.params.id}.m3u`);

Reviewer: ```suggestion
  // Liquidsoap will use the file extension from the `Content-Disposition` to guess the
  // playlist format
  res.set(`Content-Disposition: attachment; filename="playlist-${req.params.id}.m3u`);
  
  // Otherwise, it will try to guess the file extension from the playlist mime-type.
  res.set('Content-Type', 'audio/x-mpegurl')
```

and some programming magic. You can check the code source for `playlist.reloadable` in our standard library
for a detailed example.

### Remote M3U playlist example
Here is an example of a m3u playlist being read from nodejs/express .

liquidsoap script:
```liquidsoap
#!/usr/local/bin/liquidsoap

p = playlist(reload=10, "http://localhost:8080/radio/playlists/0/playlist.m3u")
```

nodejs/express app:
```js
import express from "express";
const app = express();
app.get('/radio/playlists/:id/playlist.m3u', async (req, res) => {
  const playlist = ["/media/foo.mp3", "/media/bar.mp3"]

>   // Liquidsoap will use the file extension from the `Content-Disposition` header to guess

Reviewer: ```suggestion

  // Liquidsoap will use the file extension from the `Content-Disposition` header to guess
```

and some programming magic. You can check the code source for `playlist.reloadable` in our standard library
for a detailed example.

### Remote M3U playlist example
Here is an example of a m3u playlist being read from nodejs/express .

liquidsoap script:
```liquidsoap
#!/usr/local/bin/liquidsoap

p = playlist(reload=10, "http://localhost:8080/radio/playlists/0/playlist.m3u")
```

nodejs/express app:
```js
import express from "express";
const app = express();
app.get('/radio/playlists/:id/playlist.m3u', async (req, res) => {
  const playlist = ["/media/foo.mp3", "/media/bar.mp3"]

  // Liquidsoap will use the file extension from the `Content-Disposition` header to guess
  // the playlist format

>   res.set(`Content-Disposition: attachment; filename="playlist-${req.params.id}.m3u`);

Reviewer: ```suggestion
  res.set(`Content-Disposition: attachment; filename="playlist-${req.params.id}.m3u"`);
```
